### PR TITLE
(draft) feat: use web workers for placeholder hash calculation

### PR DESCRIPTION
**Pull request by felixranesberger**
### ❓ Type of change

- [x] 📖 Documentation (updates to the documentation, readme or JSDoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [x] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by johannschopplich** [comment]
Thanks for the effort. There are many moving parts. I'm not sure what implications the async API has yet, since every framework integration depends on a sync API. Thus, it is a breaking change. I will take a deep dive later.

**Comment by felixranesberger** [comment]
Currently I'm working on the web worker support for unbuild, because since a library is shipped, we have to add a rollup web worker plugin to the unbuild configuration. 

Regarding the breaking async changes: I think it should mostly be fine, but I'll have a closer look, when I have time to work on the pull request further. Thanks for your first feedback 👍

**Comment by johannschopplich** [comment]
Some food for thought on this PR:
- Although I like the idea of using workers, a performance test would be great to get an idea of how big the gain really is.
- We need to make sure the web workers work in Node.js, browsers and worker environments. Unlazy is specifically written to provide SSR support out of the box, which we need to maintain. 
  - This is probably related to your unlazy web worker support implementation. Good luck with it!
- Please also update the framework integrations, as the async placeholder generation would also require async watchers.

The bottom line is: I'm OK with the current synchronous implementation. I can only provide limited resources on the worker technology shift. But I'm still open to that one! It just has to be implemented airtight to not break user land.

**Comment by felixranesberger** [comment]
I had some fun implementing the `rollup plugin` for `web workers` today, but to no success. 
Switching from `unbuild` to `vite` for building the core package would probably fix the issue,
since `vite` offers built in support for web workers. 

I'm not a huge fan, but I think for the moment it's the only way, I'm getting the web worker support to work. Will have a look at it maybe tomorrow, to see what actual performance benefits the web worker approach brings to the table 👍

**Comment by chrisspiegl** [comment]
This looks very promising and I appreciate the work being done on this 👍. Thank you.

**Comment by felixranesberger** [comment]
Closing this pull request, since I haven't made any progress since quite some time now. 
I'm also not in need of this feature right now, so I had no time to work on it. 
Feel free to use my state and develop it further 👍

**Comment by chrisspiegl** [comment]
I understand the reasons for closing this PR… but to be honest, it would be a great addition for slower devices especially. I have no experience coding workers… but the idea to offload the processing sounds like a solid plan.

**Comment by felixranesberger** [comment]
The main issue is that web workers need their own file and you'll need to make sure, that this file is copied and referenced to the correct place in multiple possible environments. Also using the library with Nuxt would probably mean, that we'll have to look at how Nuxt handles web workers. This could be an issue, because code can be executed on both the server and client. 

Writing the actual web worker logic should be pretty easy though.
Have you gained any experience with web workers in combination with libraries Johann?

I would be open to implement the web worker logic, even if I don't directly use the library anymore :)

**Comment by chrisspiegl** [comment]
I understand this better now, thank you, @felixranesberger.

I am not sure if it's possible to implement this so universally that one would just need a few lines to be added to the web worker; but I also understand that this is a much more complex approach in comparison to a "simple" module add in Nuxt.
